File: ocaml/change/change.ml
open Base

let make_change ~target ~coins =
  if target < 0 then Error "target can't be negative"
  else
    let rec make_change0 target ?(results = []) = function
      | [] ->
          results |> List.filter_opt
          |> List.sort ~compare:(fun a b ->
                 Int.compare (List.length a) (List.length b))
          |> List.hd
          |> Result.of_option ~error:"can't make target with given coins"
      | coins ->
          let rec compute_change target coins result =
            if target = 0 then Some result
            else
              match coins with
              | [] -> None
              | coin :: coins_left ->
                  if target >= coin then
                    match
                      compute_change (target - coin) coins (coin :: result)
                    with
                    | None ->
                        compute_change (target - coin) coins_left
                          (coin :: result)
                    | Some greedy_result -> Some greedy_result
                  else compute_change target coins_left result
          in
          make_change0 target (List.tl_exn coins)
            ~results:(compute_change target coins [] :: results)
    in
    coins |> List.sort ~compare:Int.compare |> List.rev |> make_change0 target

File: python/change/change.py
def find_fewest_coins(coins: list[int], target: int) -> list[int]:
    if target < 0:
        raise ValueError("target can't be negative")

    solutions: list[dict[int, int] | None] = [{}]
    reversed_coins = sorted((coin for coin in coins if coin <= target), reverse=True)
    for partial_target in range(1, target + 1):
        min_solution = None
        for coin in reversed_coins:
            if coin <= partial_target:
                partial_solution = solutions[partial_target - coin]

                if partial_solution is not None and (
                    min_solution is None
                    or sum(partial_solution.values()) + 1 < sum(min_solution.values())
                ):
                    min_solution = partial_solution.copy()

                    if coin in min_solution:
                        min_solution[coin] += 1
                    else:
                        min_solution[coin] = 1

        solutions.append(min_solution)

    solution = solutions[target]

    if solution is None:
        raise ValueError("can't make target with given coins")

    result: list[int] = []
    for coin in reversed_coins:
        if coin in solution:
            result.extend([coin] * solution[coin])

    result.sort()

    return result
